core/src/main/java/com/linecorp/armeria/common/HttpRequestBuilder.java

```diff
@@ -0,0 +1,371 @@
+        requireNonNull(pathParams, "pathParams");
+        this.pathParams.putAll(pathParams);
+        return this;
+    }
+
+    /**
+     * Sets a query param for this request. For example:
+     * <pre>{@code
+     * HttpRequest.get("/endpoint")
+     *            .queryParam("foo", "bar")
+     *            .build(); // GET `/endpoint?foo=bar`
+     * }</pre>
+     */
+    public HttpRequestBuilder queryParam(String name, String value) {
+        requireNonNull(name, "name");
+        requireNonNull(value, "value");
+        queryParams = queryParams.toBuilder().add(name, value).build();
+        return this;
+    }
+
+    /**
+     * Sets multiple query params for this request. For example:
+     * <pre>{@code
+     * HttpRequest.get("/endpoint")
+     *            .queryParams(QueryParams.of("from", "foo", "limit", 10))
+     *            .build(); // GET `/endpoint?from=foo&limit=10`
+     * }</pre>
+     * @see QueryParams
+     */
+    public HttpRequestBuilder queryParams(QueryParams queryParams) {
+        requireNonNull(queryParams, "queryParams");
+        final QueryParamsBuilder queryParamsBuilder = this.queryParams.toBuilder();
+        for (String name : queryParams.names()) {
+            queryParamsBuilder.add(name, requireNonNull(queryParams.get(name), "query"));
+        }
+        this.queryParams = queryParamsBuilder.build();
+        return this;
+    }
+
+    /**
+     * Sets a cookie for this request. For example:
+     * <pre>{@code
+     * HttpRequest.get("/")
+     *            .cookie(Cookie.of("cookie", "foo"))
+     *            .build();
+     * }</pre>
+     * @see Cookie
+     */
+    public HttpRequestBuilder cookie(Cookie cookie) {
+        requireNonNull(cookie, "cookie");
+        cookies.add(cookie);
+        return this;
+    }
+
+    /**
+     * Sets multiple cookies for this request. For example:
+     * <pre>{@code
+     * HttpRequest.get("/")
+     *            .cookies(Cookies.of(Cookie.of("cookie1", "foo"),
+     *                                Cookie.of("cookie2", "bar")))
+     *            .build();
+     * }</pre>
+     * @see Cookies
+     */
+    public HttpRequestBuilder cookies(Cookies cookies) {
+        requireNonNull(cookies, "cookies");
+        this.cookies.addAll(cookies);
+        return this;
+    }
+
+    /**
+     * Makes this request a streaming request.
+     */
+    public HttpRequestBuilder streaming() {
+        this.streaming = true;
+        return this;
+    }
+
+    /**
+     * Creates a new {@link HttpRequest}.
+     */
+    public HttpRequest build() {
+        if (!pathParams.isEmpty()) {
+            path = replacePathParams();
+        }
+        if (!queryParams.isEmpty()) {
+            path += "?" + queryParams.toQueryString();
+        }
+        final RequestHeaders requestHeaders = buildRequestHeader();
+        if (streaming) {
+            return new DefaultHttpRequest(requestHeaders);
```

Comment:
I'm not sure if the `streaming` property is a good idea because `build()` returns an `HttpRequest` rather than `HttpRequestWriter`. How about focusing on providing the API for building non-streaming requests? We can later add a method like `content(..., Publisher<HttpData>)`.